### PR TITLE
♻️(api) factorize video's indicators

### PR DESCRIPTION
**Pull request by lebaudantoine**
## Purpose

Reduce code's duplication between video's indicators.

## Proposal

Abstract the views and downloads indicators into a more modular general indicator. 

- [x] Offer a reusable indicator class.
- [x] Provide additional feature to the indicator using mixin.
- [x] Update documentation
- [x] Update API routes
- [x] Update tests

### CODE REVIEW

File: src/api/core/warren/base_indicator.py [hunk 1]
```diff
@@ -64,9 +82,13 @@ def get_lrs_query(self) -> LRSQuery:
...
 
     @abstractmethod
-    def fetch_statements(self) -> List:
+    def fetch_statements(self):
         """Executes the LRS query to obtain statements required for the indicator."""
 
+    @abstractmethod
+    def filter_statements(self):
+        """Filters the statements required for the indicator."""
+
```
Comment (lebaudantoine) [inline src/api/core/warren/base_indicator.py:91]: Should it be an `@abstractmethod` from `BaseIndicator` or added/provided with inheritance of a `FilterMixin`? I am open to discuss it.

File: src/api/plugins/video/warren_video/api.py [hunk 2]
```diff
@@ -28,26 +30,39 @@ async def views(
...
+    query_parameters = VideoQueryParameters(
+        video_id=video_id,
+        verb_id=verb_id,
+        date_range=DatetimeRange(since=filters.since, until=filters.until),
+    )
+    indicator = DailyVideoEventIndicator(
+        client=lrs_client,
+        query_parameters=query_parameters,
+        remove_duplicate_actors=unique,
+        statements_filter=statements_filter,
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:56]: Should we group those arguments into a single structure, such as `filter_parameters`, to show that they're all used in the same method?

File: src/api/plugins/video/warren_video/api.py [hunk 3]
```diff
@@ -28,26 +30,39 @@ async def views(
...
-        indicator = DailyVideoViews(**indicator_kwargs)
+        # todo - extract this filter definition?
+        def filter_statements_on_duration(row):
+            return (
+                row[f"result.extensions.{RESULT_EXTENSION_TIME}"]
+                >= video_plugin_settings.VIEWS_COUNT_TIME_THRESHOLD
+            )
+
+        verb_id = PlayedVerb().id
+        statements_filter = filter_statements_on_duration
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:45]: It feels like the route deeply change its behavior based on a `if/else`

File: src/api/plugins/video/warren_video/api.py [hunk 4]
```diff
@@ -28,26 +30,39 @@ async def views(
...
+        query_parameters=query_parameters,
+        remove_duplicate_actors=unique,
+        statements_filter=statements_filter,
+    )
     try:
-        response = Response[DailyCounts](
-            status=StatusEnum.SUCCESS, content=indicator.compute()
-        )
+        result = indicator.compute()
+        response = Response[DailyCounts](status=StatusEnum.SUCCESS, content=result)
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:60]: could be condensed back? WDYT ?

File: src/api/plugins/video/warren_video/api.py [hunk 5]
```diff
@@ -58,19 +73,22 @@ async def downloads(
...
+        client=lrs_client,
+        query_parameters=query_parameters,
+        remove_duplicate_actors=unique,
     )
     try:
-        response = Response[DailyCounts](
-            status=StatusEnum.SUCCESS, content=indicator.compute()
-        )
+        result = indicator.compute()
+        response = Response[DailyCounts](status=StatusEnum.SUCCESS, content=result)
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:88]: ditto.

File: src/api/plugins/video/warren_video/indicators.py [hunk 6]
```diff
@@ -1,269 +1,103 @@
...
-from warren.models import XAPI_STATEMENT, DailyCounts
-from warren_video.conf import settings as video_plugin_settings
+from warren.base_indicator import BaseIndicator, PreprocessMixin
+from warren.models import DailyCounts
+from warren_video.models import VideoQueryParameters
 
 
-class DailyVideoViews(BaseIndicator):
-    """Daily Video Views indicator.
+class DailyVideoEventIndicator(BaseIndicator, PreprocessMixin):
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/indicators.py:11]: is `Event` the right naming? Anything closer to the xAPI semantic?

File: src/api/plugins/video/warren_video/indicators.py [hunk 7]
```diff
@@ -1,269 +1,103 @@
...
-
-        # Calculate the total number of events
-        indicator.total_count = len(filtered_view_duration.index)
-        indicator.count_by_date = count_by_date.to_dict("records")
-
-        return indicator
+        if self.statements_filter:
+            self.statements = self.statements[
+                self.statements.apply(self.statements_filter, axis=1)
+            ]
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/indicators.py:70]: Does this provide sufficient modularity? 
Should we have another indicator class defined, that calls the parent method and surcharges it to handle some custom filters?

**Review by lebaudantoine — COMMENTED** [review]
Let's discuss some parts of this refactoring ! I would appreciate some feedbacks on the `filter_statements` parts.

File: src/api/core/warren/base_indicator.py (hunk 1 above)
Comment (jmaupetit) [reply src/api/core/warren/base_indicator.py:91]: An abstract method from the `BaseIndicator` seems fine to me.

File: src/api/plugins/video/warren_video/models.py [hunk 8]
```diff
@@ -1 +1,12 @@
...
+from pydantic.main import BaseModel
+from warren.filters import DatetimeRange
+
+
+class VideoQueryParameters(BaseModel):
+    """Base model to represent an LRS query for fetching video statements."""
+
+    video_id: str
+    verb_id: str
+    date_range: DatetimeRange
```
Comment (jmaupetit) [inline src/api/plugins/video/warren_video/models.py:12]: If not used in the API request signature, it seems fragile to me.

**Review by jmaupetit — COMMENTED** [review]
Interesting pattern! Don't miss the point: API views should be as dumb as possible, an indicator should be self-sufficient and explicit; it can be used in various contexts, outside from the API, such as cron task, etc.

**Review by jmaupetit — COMMENTED** [review]
It looks way better to me. Cannot wait to see all your current improvements merged!

I'll let @quitterie-lcs and @wilbrdt validate this work while I'm away. Thanks!

**Review by wilbrdt — APPROVED** [review]
That's nice and clean! I like it :100: 
Just nit-picking the docstrings imperative style but good to go! :tada:

**Pull request merged**